### PR TITLE
Adds locks and keys

### PR DESCRIPTION
**Pull request by UndeadNappist**
Made the locking proc actually function on /door. There's probably some spaghetti rendundancy that needs fixing in terms of operating doors, (open), bumpopen, and other shenanigans but I'm not doing it right now.

Added descriptions of key and lock under New() to ensure that people can check what number the keys and locks are set to.

With steel sheets in hand:
Manufacture a lock (or multiple).
Manufacture a key (or multiple).
Use the keys on the locks to set them to be the same as that lock.
Use the locks on other locks to set them to be the same.
Apply the locks to doors.
While a door is unlocked, apply a lock to set it to be the same.
If you have a key, apply a single rod to the key to manufacture an identical key.

Things that still need doing: An overlay sprite for doors with a non-ugly lock, better lock, in-hand sprites for both lock and key.

Special thanks to Rhicora and d20 folks who helped me out with this.

Added bonus: Comes with a map fix that replaces a space square with desert (Mr. Moose pointed this out).

### CODE REVIEW

**Comment by AndrewL97** [comment]
@LetterN im writing up a change that Id prefer to have implemented
essentially add_lock and check_locked procs on the lock class itself, to be called when the door is bumped

that way we can easily add locks to whatever we want, without rewriting the code blob

please no merge yet

**Comment by UndeadNappist** [comment]
Looks like /closet/ uses locked var and it's /closet/crate, so it might be doable to just rewrite mineral doors to use locked = true and make it work that way.

Want me to do that?

**Comment by AndrewL97** [comment]
No, leave the locked var as is

the only changes that should be made to the doors (machine and mineral
doors) should be a variable to hold the lock object
(/var/item/lock_construct)
and then the check_locked() and add_lock() procedure calls in the bump/open
attempts

the actual body of check_locked, etc. should be in the lock_construct class

**Comment by AndrewL97** [comment]
https://gist.github.com/AndrewL97/97823b31d0899df70a952d416363189c

@UndeadNappist does this make sense for what I want you to do?

**Comment by UndeadNappist** [comment]
> https://gist.github.com/AndrewL97/97823b31d0899df70a952d416363189c
> 
> @UndeadNappist does this make sense for what I want you to do?

The problem is adding the code that way will still make it so if someone spam-clicks on /unpowered/ the door will still force itself open because of how attackby works in /machinery/door, which is why I had to add this line of code https://github.com/BadDeathclaw/TG-Claw/pull/178/commits/2b713cfcc160a4194bee007343bcf96f6ffcf06c#diff-21a156317b49600c87cd9473e7aef34dL148 

You would need to specifically interact with /obj/machinery/door/proc/try_to_activate_door(mob/user) or something along those lines - using attackby is a recipe for making it impossible to interact with the door or, alternatively, allowing someone to spam-click their way through the door and ignore the locking mechanism entirely.

You could also add all the procs under /door/unpowered instead of just /door/ and that way they wouldn't interact with airlocks at all, but would affect every type of /unpowered/ if there was ever a decision to add more types of it that use the same procs and effects instead of going with mineral_doors (and one of the types - either /unpowered/door or mineral_doors should likely be rendered defunct and moved over for simplicity).

Mineral_door would need var/locked added to them anyway with this change.

**Comment by UndeadNappist** [comment]
I realize what you're trying to do, which is add an entirely separate, _additional_ check under lock_construct to handle bumpopen and whether or not the door is locked. I don't think it's a good idea to add additional procs and checks when there is already (unfortunately several) ways of handling it already.

**Comment by AndrewL97** [comment]
> I realize what you're trying to do, which is add an entirely separate, _additional_ check under lock_construct to handle bumpopen and whether or not the door is locked. I don't think it's a good idea to add additional procs and checks when there is already (unfortunately several) ways of handling it already.

The locked variable you are using in your checks if for door bolts, you shouldn't be using it or setting it with your padlock changes

File: config/admins.txt [hunk 1]
```diff
@@ -9,4 +9,4 @@
 # If SQL-based admin loading is enabled, admins listed here will always be loaded first		  #
 # and will override any duplicate entries in the database.									  #
 ###############################################################################################
-Gomble = Host
+Gomble = Host
```
[the remark this answers is not in the document]
Comment (UndeadNappist) [reply config/admins.txt]: https://github.com/BadDeathclaw/TG-Claw/pull/178/commits/2b713cfcc160a4194bee007343bcf96f6ffcf06c#diff-a084b794bc0759e7a6b77810e01874f2

should probably gitignore the rest of the config files, honestly

File: code/game/machinery/doors/door.dm [hunk 2]
```diff
@@ -80,7 +80,7 @@
 	return ..()
 
 /obj/machinery/door/CollidedWith(atom/movable/AM)
-	if(operating || (obj_flags & EMAGGED))
+	if(operating || (obj_flags & EMAGGED) || (locked == TRUE))
```
[the remark this answers is not in the document]
Comment (UndeadNappist) [reply code/game/machinery/doors/door.dm]: Helpful.

**Comment by UndeadNappist** [comment]
>The only thing I want you to rewrite, is WHERE the checks for correct keys, whether the lock is locked, etc. 
That is a lie.
var/locked is already handled by the object you are trying to interact with.
Procs for opening are handled by the objects I have fiddled with.
You want to add new procs for checking whether or not a door should open in an entirely different file despite everything you say. I take a dim view of this, given what is already compressed within /doors/ and /mineral_doors/.
If anything, those two things should be in the same hierarchy, but they are not.
I decline to do this, since it will add even more lines of code and checks, despite your claims otherwise.

>You can put the calls to the procs wherever you want, and wherever you feel they need to be.
Things are already exactly where they need to be.

>The code for the locks should be under the lock itself, and not written slightly differently for each location its used.
The code for locks is already in locks. The code for checking whether or not to open is in the affected object, and how to interact with items if they happen to be locked/not locked and have lock/not have lock.

>Send me a DM on the discord
I have sent several several days ago. You have failed to respond, and in point of fact, made this the new medium of communication as we proceeded to have precisely the same discussion we had over DMs earlier in a very circular fashion.

I decline and refuse to increase spaghetti by splitting procs across various kinds of objects unless I make something total modular that would call for adding locks to all /obj/. As that would cause a hideous amount of problems, I'm not doing that, either.

I no longer have the energy to continue this. There are other things that need doing.

**Pull request closed without merging**